### DISCUSSION TITLE
Question on DOF ordering

### DISCUSSION THREAD

**Discussion by jwpeterson** [libMesh/libmesh#3660] [Q&A]
_This question was emailed to me by Liu Haochen, I've copied it here so that others can see/comment. --John_

> Dear developers,
> 
> 
> 
> We'd like to consult the ordering of freedoms in a global matrix arising from multi-variable problems such as the Navier-Stokes equations, as addressed in ../examples/systems_of_equations/ex2.
> 
> 
> 
> 1) If we use a linear approximation for velocity and pressure (P1-P1), is the global matrix ordered in a node-by-node form as follows?
> 
> ![dof_ordering](https://github.com/libMesh/libmesh/assets/1775907/02e497d9-34e0-4ded-9053-c10cb0928042)
> 
> 
> 
> 
> 
> 
> 2) What is the ordering of the global matrix if we use "SECOND" order for velocity (P2-P1)? Since there are more dofs for velocity than pressure.
> 
> 
> 
> 3) For the example "../examples/systems_of_equations/ex2", we find that it fails when we use the option  ' -sub_pc_factor_mat_ordering_type rcm ', therefore we are confused with the ordering.
> 
> 
> 
> 
> 
> We are looking forward to your reply. Thank you so much!

**Comment by jwpeterson** [comment]
By default the variable ordering in libMesh is so-called "variable major", so in your example all the u components would appear first, followed by all the v components, and finally all the p components. Please note however that the variable ordering can be changed via command line argument, so one should avoid writing code which assumes any particular ordering.

> we find that it fails when we use the option ' -sub_pc_factor_mat_ordering_type rcm '

There could be many reasons for the failure, but I wouldn't initially suspect it is necessarily related to libMesh's DOF ordering. I don't think I've ever tried this particular PETSc option before, and we definitely don't have any testing for it.

**Reply by roystgnr** [reply]
Two slight corrections:

1. the DoF ordering in libMesh is currently always "rank-major" - all the DoFs owned by processor (MPI rank) M come before all those owned by processor N if N>M.  The variable-major (default) vs node-major (CLI `--node-major-dofs`) options are for the next layer of ordering, within-processor.

2. Our "variable-major" is really "variable-group-major", for efficiency.  Applications can add multiple variables (of the exact same family and order and subdomain restrictions) at the same time to create a variable group ... and unless an app sets `system.identify_variable_groups(false)`, even when adding new variables to `system` one at a time libMesh will see if they match the previous variable and will append them to that variable's group if so.  In a Taylor Hood (u,v,p) case where u and v match but p doesn't, IIRC the ordering will look something like [u0, v0, u1, v1, u2, v2, ... p0, p1, p2,...]

And not a correction, but hopefully helpful to Liu: I have played around with `-sub_pc_factor_mat_ordering_type` before, and I settled on `nd` as the most robust with ILU subpreconditioners for incompressible flow problems.  This was many years ago, but it might still be worth a try.

> one should avoid writing code which assumes any particular ordering.

I personally agree, but it's amusing that, IIRC, Ben Kirk's magnum opus, FIN-S, required the `--node-major-dofs` setting to work.  It's also potentially useful for efficiency; if you have n variables and they're all nodal + isoparametric then you can manually init your sparse matrix with a `blocksize` of n and (at least with a PETSc matrix) the sparsity pattern will take `n^2` less space.

**Reply by jwpeterson** [reply]
Indeed, I forgot about the variable grouping, thanks for the explanation Roy.

**Reply by LHappyCureall** [reply]
Thank you, Roy and John.
So I also want to know if we use node-major (CLI --node-major-dofs) options, in a Taylor Hood(U,P) case, what's the ordering...    
                             
If  we use node-major (--node-major-dofs) options and just use P1-P1(U,P) case, ordering like [u0,v0,p0, u1,v1,p1.......un,vn,pn], right?

**Reply by roystgnr** [reply]
That's right (about the P1-P1 case; I always thought "Taylor Hood" implied a velocity polynomial degree one larger than the pressure polynomial degree).

**Reply by LHappyCureall** [reply]
I'm really sorry to bother you again. For the P1-P1, because it is ordered node by node, we can use block size 3 for the matrix storage. But for P2-P1, we don't know how to set the block size. Would you please let us know about the exact ordering when using P2-P1 if using the "--node-major-dofs" option?

**Reply by jwpeterson** [reply]
> Would you please let us know about the exact ordering when using P2-P1 if using the "--node-major-dofs" option?

I think it will be like @roystgnr said: 
> In a Taylor Hood (u,v,p) case where u and v match but p doesn't, IIRC the ordering will look something like [u0, v0, u1, v1, u2, v2, ... p0, p1, p2,...]

But you should probably write a small test code that loops over the nodes and prints 
```
node->dof_number(sys_num, var_num, 0);
```
Note that not every node will have p DOFs in this case, and I think `dof_number()` throws an error if you ask for a DOF number that doesn't exist.

**Reply by roystgnr** [reply]
For P2-P1 there is no valid block size, even with `--node-major-dofs`.  Vertices will have [ui, vi, pi]; edge nodes (and if you use Q2-Q1 quads or hexes too, interiors and face nodes) will have just [ui, vi].  The ordering is consistent, but some nodes will have DoFs that others won't.  You can't even group multiple nodes into a single global block size: even if you're in serial and you ensure the node numbering is locally consistent with that, the node count won't be.  E.g. a regular NxN grid of 2N² triangles will have N²+2N+1 vertices but 3N²+2N edge nodes; asymptotically you have 9 DoFs per P2-P1 triangle pair but there's remainders at borders.

**Reply by roystgnr** [reply]
> `dof_number()` throws an error if you ask for a DOF number that doesn't exist.

It does so in `METHOD=dbg` or `devel` builds. In profiling or optimized builds (or any custom build without `NDEBUG` defined) asking for a DoF number that doesn't exist [summons nasal demons](http://www.catb.org/jargon/html/N/nasal-demons.html).

If you're not sure a variable has any components on node, you can use `node->n_comp(sys_num, var_num)` first to check.
